app/code/Magento/ImportExport/Model/Import/Source/FileParser/CompositeParserFactory.php

```diff
@@ -0,0 +1,62 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\ImportExport\Model\Import\Source\FileParser;
+
+/**
+ * File Parser Composite
+ *
+ */
+class CompositeParserFactory implements ParserFactoryInterface
+{
+    /** @var ParserFactoryInterface[] */
+    private $parserFactories;
+
+    public function __construct(array $parserFactories = [])
+    {
+        $this->parserFactories = [];
+
+        foreach ($parserFactories as $parserFactory) {
+            $this->addParserFactory($parserFactory);
+        }
+    }
+
+    public function create($path, array $options = [])
+    {
+        foreach ($this->parserFactories as $parserFactory) {
+            $parser = $this->createParserIfSupported($parserFactory, $path, $options);
+
+            if ($parser === false) {
+                continue;
+            }
+
+            return $parser;
+        }
+
+        $this->thereWasNoParserFound($path);
+    }
+
+    public function addParserFactory(ParserFactoryInterface $parserFactory)
+    {
+        $this->parserFactories[] = $parserFactory;
+    }
+
+    private function thereWasNoParserFound($path)
+    {
+        throw new UnsupportedPathException($path);
+    }
+
+    private function createParserIfSupported(ParserFactoryInterface $parserFactory, $path, array $options)
+    {
+        try {
+            $parser = $parserFactory->create($path, $options);
+        } catch (UnsupportedPathException $e) {
+            return false;
```

Comment:
The root cause of the problem described above

> this does not look like a proper Object Oriented way of polymorphic object creation.
> 
> because we instantiate in advance all the factories and just looping through all of them trying to instantiate needed parser

Because we don't use proper Object Oriented design to choose correct factory to be used and instantiate needed parser. We just look over all possible factories and try to instantiate parser with the help of each next factory in the list. So, our Exception is not actually exception (because Exception represents unexpected behavior, but here we have expected false results) that's why we have a juggling of Exception -> FALSE
Which is bad practice working with Exceptions